braingeneers/utils/memoize_s3.py

```diff
@@ -0,0 +1,145 @@
+
+        # This option is available by default but we can't accept it because
+        # there's no reasonable way to make joblib use NumpyS3Memmap.
+        self.mmap_mode = backend_options.get("mmap_mode")
+        if self.mmap_mode is not None:
+            raise ValueError("impossible to mmap on S3.")
+
+        # Don't attempt to handle local files, just use the default backend
+        # for that!
+        if not location.startswith("s3://"):
+            raise ValueError("location must be an s3:// URI")
+
+        # We don't have to check that the bucket exists because joblib
+        # performs a `list_objects()` in it, but note that this doesn't
+        # actually check whether we can write to it!
+        self.location = location
+
+        # We need a boto3 client, so create it using the endpoint which was
+        # configured in awswrangler by importing smart_open_braingeneers.
+        self.client = boto3.Session().client(
+            "s3", endpoint_url=wr.config.s3_endpoint_url
+        )
+
+
+def memoize(
+    location=None, backend="s3", ignore=None, cache_validation_callback=None, **kwargs
+):
+    """
+    Memoize a function to S3 using joblib.Memory. By default, saves to
+    `s3://braingeneersdev/$S3_USER/cache`, but this can be configured by
+    explicitly providing a cache directory.
+
+    Accepts all the same keyword arguments as `joblib.Memory`, including `backend`,
+    which can be set to "local" to recover default behavior. Also accepts the
+    keyword arguments of `joblib.Memory.cache()` and passes them on. Usage:
+
+    ```
+    from braingeneers.utils.memoize_s3 import memoize
+
+    # Cache to the default location on NRP S3.
+    @memoize
+    def foo(x):
+        return x
+
+    # Cache to a different NRP S3 location.
+    @memoize("s3://braingeneers/someplace/else/idk")
+    def bar(x):
+        return x
+
+    # Ignore some parameters when deciding which cache entry to check.
+    @memoize(ignore=["verbose"])
+    def plover(x, verbose):
+        if verbose: ...
+        return x
+    ```
+
+    If the bucket doesn't exist, an error will be raised, but if the only
+    problem is permissions, silent failure to cache may be all that occurs,
+    depending on the verbosity setting.
+
+    Another known issue is that size-based cache eviction is NOT supported,
+    and will also silently fail. This is because there is no easy way to get
+    access times out of S3, so we can't find LRU entries.
+    """
+    if callable(location):
+        # This case probably means the @memoize decorator was used without
+        # arguments, but pass the kwargs on anyway just in case.
+        return memoize(
+            backend=backend,
+            ignore=ignore,
+            cache_validation_callback=cache_validation_callback,
+            **kwargs,
+        )(location)
+
+    if location is None and backend == "s3":
+        location = f"s3://braingeneersdev/{os.environ['S3_USER']}/cache"
```

Comment:
Is "S3_USER" always guaranteed to be set?

Reply:
Also, it might be bad if user names are in the (currently relatively uncluttered) root directory.  So I'd like to suggest the default location be:

location = f"s3://braingeneersdev/cache/{os.environ['S3_USER']}"

I would in fact, also enforce that here, and raise an Exception if a user selects any directory without the prefix "s3://braingeneersdev/cache/". It may be problematic if a user selects "s3://braingeneersdev/" or "s3://braingeneersdev/ephys" as their cache dir unknowingly, and then decides to clean the cache dir.

Reply:
The reason I chose this location is actually that `braingeneersdev` has a bunch of username directories already in it. If they're not supposed to be there, I'm happy to change this to whatever default you guys think is reasonable.

![image](https://github.com/braingeneers/braingeneerspy/assets/15931271/59f7e1eb-8827-4072-930d-b3a0a1baba44)

Reply:
As for `$S3_USER`, unfortunately it's definitely not guaranteed to be set. The idea was to get a particular user's uses of these memoized functions to be saved into their own user prefix, so it's supposed to fail if the user hasn't set it, since there's no way to find out where that is in general.

Reply:
Ah, sorry about the usernames, I was looking in `s3://braingeneers` and not `s3://braingeneersdev` now I realize.

Reply:
> As for `$S3_USER`, unfortunately it's definitely not guaranteed to be set. The idea was to get a particular user's uses of these memoized functions to be saved into their own user prefix, so it's supposed to fail if the user hasn't set it, since there's no way to find out where that is in general.

If that's the case, I would recommend something like:

```
S3_USER = os.environ.get('S3_USER', 'common')
```